File: src/net/mm2d/upnp/SsdpNotifyReceiver.java
/*
 * Copyright(C) 2016 大前良介(OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.util.Log;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetSocketAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * SSDP NOTIFYを受信するクラス
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
class SsdpNotifyReceiver extends SsdpServer {
    /**
     * NOTIFY受信を受け取るリスナー。
     */
    public interface NotifyListener {
        /**
         * NOTIFY受信時にコール。
         *
         * @param message 受信したNOTFYメッセージ
         */
        void onReceiveNotify(@Nonnull SsdpRequestMessage message);
    }

    private static final String TAG = "SsdpNotifyReceiver";
    private NotifyListener mListener;

    /**
     * インスタンス作成。
     *
     * @param ni 使用するインターフェース
     */
    public SsdpNotifyReceiver(@Nonnull NetworkInterface ni) {
        super(ni, SSDP_PORT);
    }

    /**
     * NOTIFY受信リスナーを登録する。
     *
     * @param listener リスナー
     */
    public void setNotifyListener(@Nullable NotifyListener listener) {
        mListener = listener;
    }

    @Override
    protected void onReceive(@Nonnull InterfaceAddress addr, @Nonnull DatagramPacket dp) {
        // アドレス設定が間違っている場合でもマルチキャストパケットの送信はできてしまう。
        // セグメント情報が間違っており、マルチキャスト以外のやり取りができない相手からのパケットは
        // 受け取っても無駄なので破棄する。
        if (!isSameSegment(addr, dp)) {
            Log.w(TAG, "Invalid segment packet received:" + dp.getAddress().toString()
                    + " " + addr.toString());
            return;
        }
        try {
            final SsdpRequestMessage message = new SsdpRequestMessage(addr, dp);
            // M-SEARCHパケットは無視する
            if (SsdpMessage.M_SEARCH.equals(message.getMethod())) {
                return;
            }
            if (!SsdpMessage.SSDP_BYEBYE.equals(message.getNts())
                    && !message.hasValidLocation()) {
                return;
            }
            if (mListener != null) {
                mListener.onReceiveNotify(message);
            }
        } catch (final IOException ignored) {
        }
    }

    private static boolean isSameSegment(@Nonnull InterfaceAddress ifa,
            @Nonnull DatagramPacket dp) {
        final InetSocketAddress sa = (InetSocketAddress) dp.getSocketAddress();
        final byte[] a = ifa.getAddress().getAddress();
        final byte[] b = sa.getAddress().getAddress();
        final int pref = ifa.getNetworkPrefixLength();
        final int bytes = pref / 8;
        final int bits = pref % 8;
        for (int i = 0; i < bytes; i++) {
            if (a[i] != b[i]) {
                return false;
            }
        }
        if (bits != 0) {
            final byte mask = (byte) (0xff << (8 - bits));
            if ((a[bytes] & mask) != (b[bytes] & mask)) {
                return false;
            }
        }
        return true;
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/cp/internal/server/SsdpNotifyReceiver.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.cp.internal.server

import net.mm2d.upnp.common.log.Logger
import net.mm2d.upnp.common.Http
import net.mm2d.upnp.common.ServerConst
import net.mm2d.upnp.common.SsdpMessage
import net.mm2d.upnp.common.internal.message.SsdpRequest
import net.mm2d.upnp.common.internal.server.*
import net.mm2d.upnp.common.internal.thread.TaskExecutors
import java.io.IOException
import java.net.InetAddress
import java.net.InterfaceAddress
import java.net.NetworkInterface

/**
 * Receiver for SSDP NOTIFY
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal class SsdpNotifyReceiver(
    private val delegate: SsdpServerDelegate
) : SsdpServer by delegate {
    private var listener: ((SsdpMessage) -> Unit)? = null
    private var segmentCheckEnabled: Boolean = true
    private var shouldNotAccept: SsdpMessage.() -> Boolean = { false }
    // VisibleForTesting
    internal val interfaceAddress: InterfaceAddress
        get() = delegate.interfaceAddress

    constructor(
        taskExecutors: TaskExecutors,
        address: Address,
        ni: NetworkInterface
    ) : this(SsdpServerDelegate(taskExecutors, address, ni, ServerConst.SSDP_PORT)) {
        delegate.setReceiver(::onReceive)
    }

    fun setSegmentCheckEnabled(enabled: Boolean) {
        segmentCheckEnabled = enabled
    }

    fun setNotifyListener(listener: ((SsdpMessage) -> Unit)?) {
        this.listener = listener
    }

    fun setFilter(predicate: (SsdpMessage) -> Boolean) {
        shouldNotAccept = { !predicate(this) }
    }

    // VisibleForTesting
    @Suppress("UNUSED_PARAMETER")
    internal fun onReceive(sourceAddress: InetAddress, sourcePort: Int, data: ByteArray, length: Int) {
        val listener = listener ?: return
        if (sourceAddress.isInvalidAddress(delegate.address, interfaceAddress, segmentCheckEnabled)) {
            return
        }
        try {
            val message = createSsdpRequestMessage(data, length)
            Logger.v { "receive ssdp multicast message from $sourceAddress in ${delegate.getLocalAddress()}:\n$message" }

            if (message.shouldNotAccept()) return
            // receive only Notify method
            if (message.getMethod() != Http.NOTIFY) return
            if (message.isNotUpnp()) return
            // ByeBye accepts it regardless of address problems because it does not communicate
            if (message.nts != SsdpMessage.SSDP_BYEBYE && message.hasInvalidLocation(sourceAddress)) return

            listener.invoke(message)
        } catch (ignored: IOException) {
        }
    }

    @Throws(IOException::class)
    fun createSsdpRequestMessage(data: ByteArray, length: Int): SsdpRequest =
        SsdpRequest.create(delegate.getLocalAddress(), data, length)
}
